pkg/model/resource/options.go

```diff
@@ -0,0 +1,259 @@
+	if opts.Kind != flect.Pascalize(opts.Kind) {
+		return fmt.Errorf("kind must be PascalCase (expected %s was %s)", flect.Pascalize(opts.Kind), opts.Kind)
+	}
+
+	// TODO: validate plural strings if provided
+
+	return nil
+}
+
+func (opts *Options) GVK() config.GVK {
+	return config.GVK{
+		Group:   opts.Group,
+		Version: opts.Version,
+		Kind:    opts.Kind,
+	}
+}
+
+// safeImport returns a cleaned version of the provided string that can be used for imports
+func (opts *Options) safeImport(unsafe string) string {
+	safe := unsafe
+
+	// Remove dashes and dots
+	safe = strings.Replace(safe, "-", "", -1)
+	safe = strings.Replace(safe, ".", "", -1)
+
+	return safe
+}
+
+func (opts *Options) NewV1Resource(c *config.Config, doResource bool) *Resource {
+	res := opts.newResource()
+
+	// NOTE: while directories can have "-" and ".", v1 needs that directory to be a Go package
+	//       and that is the reason why we remove them for the directory
+	pkg := path.Join(c.Repo, "pkg", "apis", res.GroupPackageName, opts.Version)
+	domain := c.Domain
+
+	// pkg and domain may need to be changed in case we are referring to a builtin core resource:
+	//  - Check if we are scaffolding the resource now           => project resource
+	//  - Check if we already scaffolded the resource            => project resource
+	//  - Check if the resource group is a well-known core group => builtin core resource
+	//  - In any other case, default to                          => project resource
+	// TODO: need to support '--resource-pkg-path' flag for specifying resourcePath
+	if !doResource {
+		file := filepath.Join("pkg", "apis", res.GroupPackageName, opts.Version,
+			fmt.Sprintf("%s_types.go", strings.ToLower(opts.Kind)))
+		if _, err := os.Stat(file); os.IsNotExist(err) {
+			if coreDomain, found := v1CoreGroups[opts.Group]; found {
+				pkg = path.Join("k8s.io", "api", opts.Group, opts.Version)
+				domain = coreDomain
+			}
+		}
+	}
+
+	res.Package = pkg
+	res.Domain = opts.Group
+	if domain != "" {
+		res.Domain += "." + domain
+	}
+
+	res.CreateExampleReconcileBody = opts.CreateExampleReconcileBody
+
+	return res
+}
+
+func (opts *Options) NewResource(c *config.Config, doResource bool) *Resource {
+	res := opts.newResource()
+
+	pkg := path.Join(c.Repo, "api", opts.Version)
+	if c.MultiGroup {
+		pkg = path.Join(c.Repo, "apis", res.Group, opts.Version)
+	}
+	domain := c.Domain
+
+	// pkg and domain may need to be changed in case we are referring to a builtin core resource:
+	//  - Check if we are scaffolding the resource now           => project resource
+	//  - Check if we already scaffolded the resource            => project resource
+	//  - Check if the resource group is a well-known core group => builtin core resource
+	//  - In any other case, default to                          => project resource
+	// TODO: need to support '--resource-pkg-path' flag for specifying resourcePath
+	if !doResource {
+		if !c.HasResource(opts.GVK()) {
```

Comment:
V2 has an optimization, using the tracked resources instead of looking for the file if `doController` wasn't set. Unfortunately V1 doesn't track resources in the config file.